unison-share-api/src/Unison/Server/Endpoints/NamespaceListing.hs

```diff
@@ -154,61 +159,92 @@ serve ::
   Maybe NamespaceFQN ->
   Maybe NamespaceFQN ->
   Backend.Backend IO NamespaceListing
-serve codebase maySBH mayRelativeTo mayNamespaceName =
-  let -- Various helpers
-      errFromEither f = either (throwError . f) pure
-
-      parsePath :: String -> Backend IO Path.Path'
-      parsePath p = errFromEither (`Backend.BadNamespace` p) $ Path.parsePath' p
-
-      findShallow ::
-        ( V2Branch.Branch IO ->
-          IO [Backend.ShallowListEntry Symbol Ann]
-        )
-      findShallow branch = Backend.lsShallowBranch codebase branch
-
-      makeNamespaceListing ::
-        ( PPE.PrettyPrintEnv ->
-          UnisonName ->
-          UnisonHash ->
-          [Backend.ShallowListEntry Symbol a] ->
-          Backend IO NamespaceListing
-        )
-      makeNamespaceListing ppe fqn hash entries =
-        pure . NamespaceListing fqn hash $
-          fmap
-            (backendListEntryToNamespaceObject ppe Nothing)
-            entries
-
-      -- Lookup paths, root and listing and construct response
-      namespaceListing :: Backend IO NamespaceListing
-      namespaceListing = do
-        -- Relative and Listing Path resolution
-        --
-        -- The full listing path is a combination of the relativeToPath (prefix) and the namespace path
-        --
-        -- For example:
-        --            "base.List"    <>    "Nonempty"
-        --                ↑                    ↑
-        --         relativeToPath        namespacePath
-        --
-        -- resulting in "base.List.map" which we can use via the root branch (usually the codebase hash)
-        -- to look up the namespace listing and present shallow name, so that the
-        -- definition "base.List.Nonempty.map", simple has the name "map"
-        --
-        relativeToPath' <- (parsePath . Text.unpack) $ fromMaybe "." mayRelativeTo
-        namespacePath' <- (parsePath . Text.unpack) $ fromMaybe "." mayNamespaceName
-
-        let path = Path.fromPath' relativeToPath' <> Path.fromPath' namespacePath'
-        let path' = Path.toPath' path
-
-        mayRootHash <- traverse (Backend.expandShortBranchHash codebase) maySBH
-        listingCausal <- Backend.getShallowCausalAtPathFromRootHash codebase mayRootHash path
-        listingBranch <- liftIO $ V2Causal.value listingCausal
-        shallowPPE <- liftIO $ Backend.shallowPPE codebase listingBranch
-        let listingFQN = Path.toText . Path.unabsolute . either id (Path.Absolute . Path.unrelative) $ Path.unPath' path'
-        let listingHash = v2CausalBranchToUnisonHash listingCausal
-        listingEntries <- lift (findShallow listingBranch)
-
-        makeNamespaceListing shallowPPE listingFQN listingHash listingEntries
-   in namespaceListing
+serve codebase maySBH mayRelativeTo mayNamespaceName = do
+  useIndex <- asks Backend.useNamesIndex
+  mayRootHash <- traverse (Backend.expandShortBranchHash codebase) maySBH
```

Comment:
Can there be multiple maySBH matches, and what happens then?

Reply:
We have an error case for that: https://github.com/unisonweb/unison/blob/f32945d92b5ce486e8396729890fefa63ae75442/unison-share-api/src/Unison/Server/Backend.hs#L778